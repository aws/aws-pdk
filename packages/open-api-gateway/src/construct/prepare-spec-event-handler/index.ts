/*********************************************************************************************************************
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License").
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ******************************************************************************************************************** */
import * as crypto from "crypto";
import { S3 } from "aws-sdk"; // eslint-disable-line
import { prepareApiSpec, PrepareApiSpecOptions } from "./prepare-spec";

/**
 * Represents an object location in an s3 bucket
 */
export interface S3Location {
  /**
   * The bucket in which the object resides
   */
  readonly bucket: string;
  /**
   * The object key
   */
  readonly key: string;
}

/**
 * Properties required to prepare the api specification with the given integrations, authorizers, etc
 */
export interface PrepareApiSpecCustomResourceProperties
  extends PrepareApiSpecOptions {
  /**
   * The location from which to read the spec to prepare
   */
  readonly inputSpecLocation: S3Location;
  /**
   * The location to write the prepared spec. Note that the key is used as a prefix and the output location will
   * include a hash.
   */
  readonly outputSpecLocation: S3Location;
}

/**
 * Cloudformation event type for custom resource
 */
interface OnEventRequest {
  /**
   * The type of cloudformation request
   */
  readonly RequestType: "Create" | "Update" | "Delete";
  /**
   * Physical resource id of the custom resource
   */
  readonly PhysicalResourceId?: string;
  /**
   * Properties for preparing the api
   */
  readonly ResourceProperties: PrepareApiSpecCustomResourceProperties;
}

/**
 * Custom resource response
 */
interface OnEventResponse {
  /**
   * Physical resource id of the custom resource
   */
  readonly PhysicalResourceId: string;
  /**
   * Status of the custom resource
   */
  readonly Status: "SUCCESS" | "FAILED";
  /**
   * Data returned by the custom resource
   */
  readonly Data?: {
    /**
     * The key for the output spec in the output bucket
     */
    readonly outputSpecKey: string;
  };
}

const s3 = new S3();

/**
 * Prepare the api spec for API Gateway
 * @param inputSpecLocation location of the specification to prepare
 * @param outputSpecLocation location to write the prepared spec to
 * @param options integrations, authorizers etc to apply
 * @return the output location of the prepared spec
 */
const prepare = async ({
  inputSpecLocation,
  outputSpecLocation,
  ...options
}: PrepareApiSpecCustomResourceProperties): Promise<S3Location> => {
  // Read the spec from the s3 input location
  const inputSpec = JSON.parse(
    (
      await s3
        .getObject({
          Bucket: inputSpecLocation.bucket,
          Key: inputSpecLocation.key,
        })
        .promise()
    ).Body!.toString("utf-8")
  );

  // Prepare the spec
  const preparedSpec = prepareApiSpec(inputSpec, options);
  const preparedSpecHash = crypto
    .createHash("sha256")
    .update(JSON.stringify(preparedSpec))
    .digest("hex");

  const outputLocation: S3Location = {
    bucket: outputSpecLocation.bucket,
    key: `${outputSpecLocation.key}/${preparedSpecHash}.json`,
  };

  // Write the spec to the s3 output location
  await s3
    .putObject({
      Bucket: outputLocation.bucket,
      Key: outputLocation.key,
      Body: JSON.stringify(preparedSpec),
    })
    .promise();

  return outputLocation;
};

exports.handler = async (event: OnEventRequest): Promise<OnEventResponse> => {
  switch (event.RequestType) {
    case "Create":
    case "Update":
      // Prepare the spec on create
      const outputLocation = await prepare(event.ResourceProperties);
      return {
        PhysicalResourceId: outputLocation.key,
        Status: "SUCCESS",
        Data: {
          outputSpecKey: outputLocation.key,
        },
      };
    case "Delete":
    // Nothing to do for delete
    default:
      break;
  }

  return {
    PhysicalResourceId: event.PhysicalResourceId!,
    Status: "SUCCESS",
  };
};
