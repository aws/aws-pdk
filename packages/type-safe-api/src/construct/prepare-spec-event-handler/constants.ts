/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
/**
 * HTTP Methods supported by Open API v3
 */
export enum HttpMethods {
  GET = "get",
  PUT = "put",
  POST = "post",
  DELETE = "delete",
  OPTIONS = "options",
  HEAD = "head",
  PATCH = "patch",
  TRACE = "trace",
}

/**
 * Default authorizer identifiers
 */
export enum DefaultAuthorizerIds {
  NONE = "none",
  IAM = "aws.auth.sigv4",
  API_KEY = "api_key",
}
