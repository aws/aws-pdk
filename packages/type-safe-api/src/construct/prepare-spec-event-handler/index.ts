/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as crypto from "crypto";
import { // eslint-disable-line
  S3Client,
  GetObjectCommand,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import { prepareApiSpec, PrepareApiSpecOptions } from "./prepare-spec";

/**
 * Represents an object location in an s3 bucket
 */
export interface S3Location {
  /**
   * The bucket in which the object resides
   */
  readonly bucket: string;
  /**
   * The object key
   */
  readonly key: string;
}

/**
 * Properties required to prepare the api specification with the given integrations, authorizers, etc
 */
export interface PrepareApiSpecCustomResourceProperties {
  /**
   * The location from which to read the spec to prepare
   */
  readonly inputSpecLocation: S3Location;
  /**
   * The location from which to read the configuration for the input spec (integrations, cors, authorizers, etc)
   */
  readonly inputConfigLocation: S3Location;
  /**
   * The location to write the prepared spec. Note that the key is used as a prefix and the output location will
   * include a hash.
   */
  readonly outputSpecLocation: S3Location;
}

/**
 * Cloudformation event type for custom resource
 */
interface OnEventRequest {
  /**
   * The type of cloudformation request
   */
  readonly RequestType: "Create" | "Update" | "Delete";
  /**
   * Physical resource id of the custom resource
   */
  readonly PhysicalResourceId?: string;
  /**
   * Properties for preparing the api
   */
  readonly ResourceProperties: PrepareApiSpecCustomResourceProperties;
}

/**
 * Custom resource response
 */
interface OnEventResponse {
  /**
   * Physical resource id of the custom resource
   */
  readonly PhysicalResourceId: string;
  /**
   * Status of the custom resource
   */
  readonly Status: "SUCCESS" | "FAILED";
  /**
   * Data returned by the custom resource
   */
  readonly Data?: {
    /**
     * The key for the output spec in the output bucket
     */
    readonly outputSpecKey: string;
  };
}

const s3 = new S3Client({
  customUserAgent: `aws-pdk/type-safe-api/prepare-spec`,
});

/**
 * Prepare the api spec for API Gateway
 * @param inputSpecLocation location of the specification to prepare
 * @param inputConfigLocation location of the configuration for the input spec (integrations, cors, authorizers, etc)
 * @param outputSpecLocation location to write the prepared spec to
 * @return the output location of the prepared spec
 */
const prepare = async ({
  inputSpecLocation,
  inputConfigLocation,
  outputSpecLocation,
}: PrepareApiSpecCustomResourceProperties): Promise<S3Location> => {
  // Read the spec from the s3 input location
  const inputSpec = JSON.parse(
    await (
      await s3.send(
        new GetObjectCommand({
          Bucket: inputSpecLocation.bucket,
          Key: inputSpecLocation.key,
        })
      )
    ).Body!.transformToString("utf-8")
  );

  const inputConfig: PrepareApiSpecOptions = JSON.parse(
    await (
      await s3.send(
        new GetObjectCommand({
          Bucket: inputConfigLocation.bucket,
          Key: inputConfigLocation.key,
        })
      )
    ).Body!.transformToString("utf-8")
  );

  // Prepare the spec
  const preparedSpec = prepareApiSpec(inputSpec, inputConfig);
  const preparedSpecHash = crypto
    .createHash("sha256")
    .update(JSON.stringify(preparedSpec))
    .digest("hex");

  const outputLocation: S3Location = {
    bucket: outputSpecLocation.bucket,
    key: `${outputSpecLocation.key}/${preparedSpecHash}.json`,
  };

  // Write the spec to the s3 output location
  await s3.send(
    new PutObjectCommand({
      Bucket: outputLocation.bucket,
      Key: outputLocation.key,
      Body: JSON.stringify(preparedSpec),
    })
  );

  return outputLocation;
};

exports.handler = async (event: OnEventRequest): Promise<OnEventResponse> => {
  switch (event.RequestType) {
    case "Create":
    case "Update":
      // Prepare the spec on create
      const outputLocation = await prepare(event.ResourceProperties);
      return {
        PhysicalResourceId: outputLocation.key,
        Status: "SUCCESS",
        Data: {
          outputSpecKey: outputLocation.key,
        },
      };
    case "Delete":
    // Nothing to do for delete
    default:
      break;
  }

  return {
    PhysicalResourceId: event.PhysicalResourceId!,
    Status: "SUCCESS",
  };
};
