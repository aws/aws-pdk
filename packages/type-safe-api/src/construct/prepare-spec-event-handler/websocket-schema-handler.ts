/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { // eslint-disable-line
  ApiGatewayV2,
  CreateModelCommandOutput,
  GetModelsCommandOutput,
  GetRoutesCommandOutput,
  Model,
  Route,
  UpdateModelCommandOutput,
} from "@aws-sdk/client-apigatewayv2";
import { // eslint-disable-line
  S3,
} from "@aws-sdk/client-s3";
import type { OpenAPIV3 } from "openapi-types";
import { S3Location } from ".";
import {
  ApiGatewaySchemaWithRefs,
  extractWebSocketSchemas,
} from "./websocket-schema";

export interface WebSocketSchemaResourceProperties {
  readonly apiId: string;
  readonly inputSpecLocation: S3Location;
  readonly serverOperationPaths: {
    [routeKey: string]: string;
  };
}

/**
 * Cloudformation event type for custom resource
 */
interface OnEventRequest {
  /**
   * The type of cloudformation request
   */
  readonly RequestType: "Create" | "Update" | "Delete";
  /**
   * Physical resource id of the custom resource
   */
  readonly PhysicalResourceId?: string;
  /**
   * Properties for preparing the websocket api models
   */
  readonly ResourceProperties: WebSocketSchemaResourceProperties;
}

interface OnEventResponse {
  /**
   * Physical resource id of the custom resource
   */
  readonly PhysicalResourceId: string;
  /**
   * Data returned by the custom resource
   */
  readonly Data?: {};
}

const BATCH_SIZE = 10;

const s3 = new S3({
  customUserAgent: `aws-pdk/type-safe-api/ws-schema`,
});
const apiGw = new ApiGatewayV2({
  customUserAgent: `aws-pdk/type-safe-api/ws-schema`,
});

/**
 * Chunk an array into sub-arrays of the given size
 */
const chunk = <T>(items: T[], size: number = BATCH_SIZE): T[][] => {
  const chunks: T[][] = [];
  for (let i = 0; i < items.length; i += size) {
    chunks.push(items.slice(i, i + size));
  }
  return chunks;
};

/**
 * Delete a batch of models
 */
const batchDeleteModels = async (
  apiId: string,
  routes: { [routeKey: string]: Route },
  models: Model[]
) => {
  for (const batch of chunk(models)) {
    await Promise.all(
      batch.map(async (m) => {
        // If there's a route for this model, and it's associated with this model, we need to first remove the association,
        // since cloudformation will delete the route later, and we're not allowed to delete a model that's still referenced
        // by a route
        if (
          routes[m.Name!] &&
          routes[m.Name!].RequestModels?.model === m.Name!
        ) {
          await apiGw.updateRoute({
            ...routes[m.Name!],
            ApiId: apiId,
            RouteId: routes[m.Name!].RouteId!,
            RequestModels: {
              model: "",
            },
            ModelSelectionExpression: undefined,
          });
        }

        await apiGw.deleteModel({
          ApiId: apiId,
          ModelId: m.ModelId!,
        });
      })
    );
  }
};

/**
 * Retrieve all models which already exist on the api
 */
const getAllModelsByRouteKey = async (
  apiId: string
): Promise<{ [routeKey: string]: Model }> => {
  let nextToken: string | undefined = undefined;
  const models: Model[] = [];
  do {
    const res: GetModelsCommandOutput = await apiGw.getModels({
      ApiId: apiId,
      NextToken: nextToken,
    });
    nextToken = res.NextToken;
    models.push(...(res.Items ?? []));
  } while (nextToken);

  // Models are named with the route key
  return Object.fromEntries(models.map((m) => [m.Name!, m]));
};

const getAllRoutesByRouteKey = async (
  apiId: string
): Promise<{ [routeKey: string]: Route }> => {
  let nextToken: string | undefined = undefined;
  const routes: Route[] = [];
  do {
    const res: GetRoutesCommandOutput = await apiGw.getRoutes({
      ApiId: apiId,
      NextToken: nextToken,
    });
    nextToken = res.NextToken;
    routes.push(...(res.Items ?? []));
  } while (nextToken);

  return Object.fromEntries(routes.map((r) => [r.RouteKey!, r]));
};

/**
 * Wrap the schema from the spec in our protocol-specific schema
 */
const wrapSchema = (schema: ApiGatewaySchemaWithRefs) => ({
  type: "object",
  properties: {
    // All inputs must have a "route" which is our route selector
    route: {
      type: "string",
    },
    // Payload references the definition
    payload: {
      $ref: "#/definitions/Payload",
    },
  },
  // When we don't have a schema, the payload can be anything, including not specified
  required: ["route", ...(schema ? ["payload"] : [])],
  definitions: {
    ...schema?.definitions,
    // The payload is of the operation schema type, or {} which means "any"
    Payload: schema?.schema ?? {},
  },
});

/**
 * Create a batch of models with the appropriate schema
 */
const batchCreateModels = async (
  apiId: string,
  routeKeys: string[],
  schemas: { [routeKey: string]: ApiGatewaySchemaWithRefs }
): Promise<{ [routeKey: string]: string }> => {
  const results: CreateModelCommandOutput[] = [];
  for (const batch of chunk(routeKeys)) {
    results.push(
      ...(await Promise.all(
        batch.map(async (routeKey) =>
          apiGw.createModel({
            ApiId: apiId,
            Name: routeKey,
            ContentType: "application/json",
            Schema: JSON.stringify(wrapSchema(schemas[routeKey])),
          })
        )
      ))
    );
  }
  return Object.fromEntries(results.map((r) => [r.Name!, r.ModelId!]));
};

/**
 * Update a batch of models with the new schema
 */
const batchUpdateModels = async (
  apiId: string,
  models: Model[],
  schemas: { [routeKey: string]: ApiGatewaySchemaWithRefs }
): Promise<{ [routeKey: string]: string }> => {
  const results: UpdateModelCommandOutput[] = [];
  for (const batch of chunk(models)) {
    results.push(
      ...(await Promise.all(
        batch.map(async (model) =>
          apiGw.updateModel({
            ApiId: apiId,
            ModelId: model.ModelId!,
            ContentType: "application/json",
            Schema: JSON.stringify(wrapSchema(schemas[model.Name!])),
          })
        )
      ))
    );
  }
  return Object.fromEntries(results.map((r) => [r.Name!, r.ModelId!]));
};

/**
 * Create or update the models
 */
const createOrUpdateModels = async (
  properties: WebSocketSchemaResourceProperties,
  routes: { [routeKey: string]: Route }
): Promise<{ [routeKey: string]: string }> => {
  const modelsByRouteKey = await getAllModelsByRouteKey(properties.apiId);
  const existingRouteKeys = new Set(Object.keys(modelsByRouteKey));
  const newRouteKeys = new Set(Object.keys(properties.serverOperationPaths));

  const deletedRouteKeys = [...existingRouteKeys].filter(
    (id) => !newRouteKeys.has(id)
  );
  console.log("Operations to delete", deletedRouteKeys);
  const addedRouteKeys = [...newRouteKeys].filter(
    (id) => !existingRouteKeys.has(id)
  );
  console.log("Operations to add", addedRouteKeys);
  const updateRouteKeys = [...newRouteKeys].filter((id) =>
    existingRouteKeys.has(id)
  );
  console.log("Operations to update", updateRouteKeys);

  // Delete all the models to delete
  await batchDeleteModels(
    properties.apiId,
    routes,
    deletedRouteKeys.map((id) => modelsByRouteKey[id])
  );

  // Load the spec
  const spec = JSON.parse(
    await (
      await s3.getObject({
        Bucket: properties.inputSpecLocation.bucket,
        Key: properties.inputSpecLocation.key,
      })
    ).Body!.transformToString("utf-8")
  ) as OpenAPIV3.Document;

  // Extract the schemas from the spec
  const schemas = extractWebSocketSchemas(
    [...addedRouteKeys, ...updateRouteKeys],
    properties.serverOperationPaths,
    spec
  );

  // Create/update the relevant models
  return {
    ...(await batchCreateModels(
      properties.apiId,
      addedRouteKeys.filter((id) => schemas[id]),
      schemas
    )),
    ...(await batchUpdateModels(
      properties.apiId,
      updateRouteKeys
        .filter((id) => schemas[id])
        .map((id) => modelsByRouteKey[id]),
      schemas
    )),
  };
};

/**
 * Delete all models
 */
const deleteModels = async (
  properties: WebSocketSchemaResourceProperties,
  routes: { [routeKey: string]: Route }
) => {
  const modelsByRouteKey = await getAllModelsByRouteKey(properties.apiId);
  await batchDeleteModels(
    properties.apiId,
    routes,
    Object.values(modelsByRouteKey)
  );
};

/**
 * Handler for creating websocket schemas
 */
exports.handler = async (event: OnEventRequest): Promise<OnEventResponse> => {
  const PhysicalResourceId =
    event.PhysicalResourceId ?? `${event.ResourceProperties.apiId}-models`;
  const routes = await getAllRoutesByRouteKey(event.ResourceProperties.apiId);
  switch (event.RequestType) {
    case "Create":
    case "Update":
      await createOrUpdateModels(event.ResourceProperties, routes);
      break;
    case "Delete":
      await deleteModels(event.ResourceProperties, routes);
      break;
    default:
      break;
  }

  return {
    PhysicalResourceId,
  };
};
