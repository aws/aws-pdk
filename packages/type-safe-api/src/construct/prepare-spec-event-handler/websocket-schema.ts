/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import type { OpenAPIV3 } from "openapi-types";
import { HttpMethods } from "./constants";
import { validatePathItem } from "./prepare-spec";

type SchemaDefinitions = { [key: string]: OpenAPIV3.SchemaObject };
export type ApiGatewaySchemaWithRefs = {
  schema: OpenAPIV3.SchemaObject;
  definitions: SchemaDefinitions;
};

const _get = (object: any | undefined, paths: string[]): any => {
  if (typeof object === "undefined") {
    return undefined;
  }
  if (paths.length === 0) {
    return object;
  }
  return _get(object[paths[0]], paths.slice(1));
};

/**
 * Return whether or not the given OpenAPI object is a reference
 */
const isRef = (obj: unknown): obj is OpenAPIV3.ReferenceObject =>
  !!obj && typeof obj === "object" && "$ref" in obj;

/**
 * Return whether a given OpenAPI object is a schema object
 */
const isSchemaObj = (obj: unknown): obj is OpenAPIV3.SchemaObject =>
  !!obj &&
  typeof obj === "object" &&
  ("type" in obj ||
    "allOf" in obj ||
    "oneOf" in obj ||
    "anyOf" in obj ||
    "not" in obj);

/**
 * Split a reference into its component parts
 * eg: #/components/schemas/Foo -> ["components", "schemas", "Foo"]
 */
const splitRef = (ref: string): string[] =>
  ref
    .slice(2)
    .split("/")
    .map((p) => p.replace(/~0/g, "~").replace(/~1/g, "/"));

/**
 * Resolve the given reference in the spec
 */
const resolveRef = (spec: OpenAPIV3.Document, ref: string): any => {
  const refParts = splitRef(ref);
  const resolved = _get(spec, refParts);
  if (!resolved) {
    throw new Error(`Unable to resolve ref ${ref} in spec`);
  }
  return resolved;
};

/**
 * Get the id of a reference to be used in the "definitions" section
 */
const getRefId = (ref: string) => splitRef(ref).join("_");

/**
 * Rewrite a reference to an API gateway model supported format
 * eg #/components/schemas/Foo -> #/definitions/components_schemas_Foo
 */
const rewriteRef = (ref: string) => `#/definitions/${getRefId(ref)}`;

/**
 * Map the given function over all refs in an OpenAPI object
 */
const mapRefs = <T>(obj: object, fn: (ref: string) => T): object => {
  // Use JSON.stringify's replacement function to simplify traversing a spec
  return JSON.parse(
    JSON.stringify(obj, (key, value) => {
      if (key === "$ref") {
        return fn(value);
      }
      return value;
    })
  );
};

/**
 * Find all references recursively starting at the given schema
 */
const findAllReferences = (
  schema: OpenAPIV3.SchemaObject,
  spec: OpenAPIV3.Document,
  seenRefs: Set<string> = new Set()
): Set<string> => {
  const newRefs = new Set<string>();
  mapRefs(schema, (ref) => newRefs.add(ref));
  const refsToSearch = [...newRefs].filter((ref) => !seenRefs.has(ref));
  const newSeenRefs = new Set([...newRefs, ...seenRefs]);
  return new Set([
    ...newSeenRefs,
    ...refsToSearch.flatMap((ref) => [
      ...findAllReferences(resolveRef(spec, ref), spec, newSeenRefs),
    ]),
  ]);
};

/**
 * Rewrite all references in the schema to instead reference #/definitions/xxx, and collect any other schemas recursively referenced by the schema
 *
 * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/models-mappings-models.html#api-gateway-request-validation-model-more-complex
 */
const rewriteSchemaReferences = (
  schema: OpenAPIV3.SchemaObject,
  spec: OpenAPIV3.Document
): ApiGatewaySchemaWithRefs => {
  // Rewrite the schema and spec, replacing references with #/definitions/xxx as per the API Gateway model format
  const rewrittenSchema = mapRefs(schema, rewriteRef) as OpenAPIV3.SchemaObject;
  const rewrittenSpec = mapRefs(spec, rewriteRef) as OpenAPIV3.Document;

  // Set of definitions that must be included for the given schema
  const definitions: SchemaDefinitions = {};

  // Find all recursive references from the schema, and add their (rewritten) schema to definitions
  [...findAllReferences(schema, spec)].forEach((ref) => {
    definitions[getRefId(ref)] = resolveRef(rewrittenSpec, ref);
  });

  return { schema: rewrittenSchema, definitions };
};

export const extractWebSocketSchema = (
  operationId: string,
  requestBody: OpenAPIV3.RequestBodyObject | OpenAPIV3.ReferenceObject,
  spec: OpenAPIV3.Document
): ApiGatewaySchemaWithRefs | undefined => {
  // Resolve the body reference, if any
  const body = isRef(requestBody)
    ? (resolveRef(spec, requestBody.$ref) as OpenAPIV3.RequestBodyObject)
    : requestBody;
  const candidateSchema = body?.content?.["application/json"]?.schema;
  if (!candidateSchema) {
    // No schema found
    return undefined;
  }

  // Resolve the top level schema reference, if any
  const rawSchema = isRef(candidateSchema)
    ? (resolveRef(spec, candidateSchema.$ref) as OpenAPIV3.SchemaObject)
    : candidateSchema;

  if (!isSchemaObj(rawSchema)) {
    throw new Error(
      `Invalid OpenAPI specification: request body for operation ${operationId} is not a valid schema`
    );
  }

  // Rewrite schema references to a format accepted by API Gateway
  return rewriteSchemaReferences(rawSchema, spec);
};

export const extractWebSocketSchemas = (
  operationIds: string[],
  serverOperationPaths: { [operationId: string]: string },
  spec: OpenAPIV3.Document
): { [operationId: string]: ApiGatewaySchemaWithRefs } => {
  const schemasByOperationId: {
    [operationId: string]: ApiGatewaySchemaWithRefs;
  } = {};
  for (const operationId of operationIds) {
    const path = serverOperationPaths[operationId];
    const pathItem = spec.paths[path]!;
    validatePathItem(path, pathItem);

    // Exactly 1 operation must be defined for each path in a websocket OpenAPI spec
    const operations = Object.values(HttpMethods).flatMap((method) =>
      pathItem[method] ? [{ ...pathItem[method], method }] : []
    );

    if (operations.length !== 1) {
      throw new Error(
        `Each path must have a single method for websocket apis. Found ${
          operations.map((o) => o.method).join(", ") || "no methods"
        }`
      );
    }

    // Extract the schema for the websocket input validation model
    if (operations[0]?.requestBody) {
      const schema = extractWebSocketSchema(
        operationId,
        operations[0].requestBody,
        spec
      );
      if (schema) {
        schemasByOperationId[operationId] = schema;
      }
    }
  }
  return schemasByOperationId;
};
