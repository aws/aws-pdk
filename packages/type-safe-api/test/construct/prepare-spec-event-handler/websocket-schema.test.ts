/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { OpenAPIV3 } from "openapi-types";
import { extractWebSocketSchemas } from "../../../src/construct/prepare-spec-event-handler/websocket-schema";

const SIMPLE_SPEC: OpenAPIV3.Document = {
  openapi: "3.0.2",
  info: {
    title: "MyService",
    version: "1.0",
    description: "A sample smithy websocket api",
  },
  paths: {
    "/SendNotification": {
      post: {
        operationId: "SendNotification",
        requestBody: {
          content: {
            "application/json": {
              schema: {
                $ref: "#/components/schemas/SendNotificationRequestContent",
              },
            },
          },
          required: true,
        },
        responses: {
          "200": {
            description: "SendNotification 200 response",
          },
        },
      },
    },
    "/SubscribeToNotifications": {
      post: {
        operationId: "SubscribeToNotifications",
        requestBody: {
          content: {
            "application/json": {
              schema: {
                $ref: "#/components/schemas/SubscribeToNotificationsRequestContent",
              },
            },
          },
          required: true,
        },
        responses: {
          "200": {
            description: "SubscribeToNotifications 200 response",
          },
        },
      },
    },
  },
  components: {
    schemas: {
      Bar: {
        type: "object",
        properties: {
          blip: {
            type: "string",
          },
        },
      },
      Foo: {
        type: "object",
        properties: {
          baz: {
            type: "integer",
            format: "int32",
          },
          bart: {
            $ref: "#/components/schemas/Bar",
          },
        },
      },
      SendNotificationRequestContent: {
        type: "object",
        properties: {
          topic: {
            type: "string",
          },
          title: {
            type: "string",
          },
          message: {
            type: "string",
          },
        },
        required: ["message", "title", "topic"],
      },
      SubscribeToNotificationsRequestContent: {
        type: "object",
        properties: {
          topic: {
            type: "string",
          },
          foo: {
            $ref: "#/components/schemas/Foo",
          },
        },
        required: ["topic"],
      },
    },
  },
};

const RECURSIVE_SPEC: OpenAPIV3.Document = {
  openapi: "3.0.2",
  info: {
    title: "MyService",
    version: "1.0",
    description: "A sample smithy websocket api",
  },
  paths: {
    "/Recursive": {
      post: {
        operationId: "Recursive",
        requestBody: {
          content: {
            "application/json": {
              schema: {
                $ref: "#/components/schemas/RecursiveRequestContent",
              },
            },
          },
        },
        responses: {
          "200": {
            description: "Recursive 200 response",
          },
        },
      },
    },
  },
  components: {
    schemas: {
      Node: {
        type: "object",
        properties: {
          value: {
            type: "string",
          },
          left: {
            $ref: "#/components/schemas/Node",
          },
          right: {
            $ref: "#/components/schemas/Node",
          },
        },
        required: ["value"],
      },
      RecursiveRequestContent: {
        type: "object",
        properties: {
          root: {
            $ref: "#/components/schemas/Node",
          },
        },
      },
    },
  },
};

describe("Extract WebSocketSchema", () => {
  it("extracts schemas from a simple spec", () => {
    expect(
      extractWebSocketSchemas(
        ["SendNotification", "SubscribeToNotifications"],
        {
          SendNotification: "/SendNotification",
          SubscribeToNotifications: "/SubscribeToNotifications",
        },
        SIMPLE_SPEC
      )
    ).toMatchSnapshot();
  });

  it("extracts schemas from a recursive spec", () => {
    expect(
      extractWebSocketSchemas(
        ["Recursive"],
        {
          Recursive: "/Recursive",
        },
        RECURSIVE_SPEC
      )
    ).toMatchSnapshot();
  });

  it("throws when multiple methods are defined for a path", () => {
    expect(() => {
      expect(
        extractWebSocketSchemas(
          ["SendNotification", "SubscribeToNotifications"],
          {
            SendNotification: "/SendNotification",
            SubscribeToNotifications: "/SubscribeToNotifications",
          },
          {
            ...SIMPLE_SPEC,
            paths: {
              ...SIMPLE_SPEC.paths,
              "/SendNotification": {
                get: {
                  ...SIMPLE_SPEC.paths["/SendNotification"]!.post!,
                },
                ...SIMPLE_SPEC.paths["/SendNotification"],
              },
            },
          }
        )
      ).toMatchSnapshot();
    }).toThrow(
      "Each path must have a single method for websocket apis. Found get, post"
    );
  });
});
