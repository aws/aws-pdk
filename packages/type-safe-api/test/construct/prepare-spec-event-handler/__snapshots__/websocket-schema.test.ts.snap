// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extract WebSocketSchema extracts schemas from a recursive spec 1`] = `
{
  "Recursive": {
    "definitions": {
      "components_schemas_Node": {
        "properties": {
          "left": {
            "$ref": "#/definitions/components_schemas_Node",
          },
          "right": {
            "$ref": "#/definitions/components_schemas_Node",
          },
          "value": {
            "type": "string",
          },
        },
        "required": [
          "value",
        ],
        "type": "object",
      },
    },
    "schema": {
      "properties": {
        "root": {
          "$ref": "#/definitions/components_schemas_Node",
        },
      },
      "type": "object",
    },
  },
}
`;

exports[`Extract WebSocketSchema extracts schemas from a simple spec 1`] = `
{
  "SendNotification": {
    "definitions": {},
    "schema": {
      "properties": {
        "message": {
          "type": "string",
        },
        "title": {
          "type": "string",
        },
        "topic": {
          "type": "string",
        },
      },
      "required": [
        "message",
        "title",
        "topic",
      ],
      "type": "object",
    },
  },
  "SubscribeToNotifications": {
    "definitions": {
      "components_schemas_Bar": {
        "properties": {
          "blip": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "components_schemas_Foo": {
        "properties": {
          "bart": {
            "$ref": "#/definitions/components_schemas_Bar",
          },
          "baz": {
            "format": "int32",
            "type": "integer",
          },
        },
        "type": "object",
      },
    },
    "schema": {
      "properties": {
        "foo": {
          "$ref": "#/definitions/components_schemas_Foo",
        },
        "topic": {
          "type": "string",
        },
      },
      "required": [
        "topic",
      ],
      "type": "object",
    },
  },
}
`;
